File: q15/03.rb
#!/usr/bin/env ruby

N = 10
STEPS = 4

dp = Array.new(N + 1, 0)
cnt = 0
dp[N] = 1

N.times do |i|
  (N + 1).times do |j|
    (1..STEPS).each do |k|
      break if k > j
      dp[j - k] += dp[j]
    end
    dp[j] = 0
  end
  cnt += dp[0] if i % 2 == 1
end

puts cnt

File: q15/03.swift
#!/usr/bin/swift

let n = 10
let steps = 4
var dp = Array(repeating: 0, count: n + 1)
var count = 0
dp[n] = 1

for i in 0..<n {
    for j in 0..<(n + 1) {
        for k in 1...steps {
            if k > j { break }
            dp[j - k] += dp[j]
        }
        dp[j] = 0
    }
    if i % 2 == 1 {
        count += dp[0]
    }
}
print(count)
